nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDBRecord.java

```diff
@@ -0,0 +1,293 @@
+                    }
+                    tablets.values().forEach(tablet -> tablet.reset());
+                }
+            }
+
+            AtomicBoolean hasRest = new AtomicBoolean(false);
+            tablets.forEach(
+                    (device, tablet) -> {
+                        if (hasRest.get() == false && tablet.rowSize != 0) {
+                            hasRest.set(true);
+                        }
+                    });
+            if (hasRest.get()) {
+                if (aligned) {
+                    session.get().insertAlignedTablets(tablets);
+                } else {
+                    session.get().insertTablets(tablets);
+                }
+            }
+        } catch (Exception e) {
+            getLogger().error("Processing failed {}", flowFile, e);
+            processSession.transfer(flowFile, REL_FAILURE);
+            return;
+        }
+        processSession.transfer(flowFile, REL_SUCCESS);
+    }
+
+    private IoTDBSchema getSchema(String property, RecordReader recordReader) throws Exception {
+        ValidationResult result;
+        IoTDBSchema schema;
+        result =
+                property != null
+                        ? validateSchemaAttribute(property)
+                        : validateSchema(recordReader.getSchema());
+
+        if (!result.getKey()) {
+            getLogger().error("The property `schema` has an error: {}", result.getValue());
+            throw new Exception();
+
+        } else {
+            if (result.getValue() != null) {
+                getLogger().warn("The property `schema` has a warn: {}", result.getValue());
+            }
+        }
+
+        schema =
+                property != null
+                        ? mapper.readValue(property, IoTDBSchema.class)
+                        : convertSchema(recordReader.getSchema());
+        return schema;
+    }
+
+    private Object getValue(Object value, TSDataType type) throws Exception {
+        if (value != null) {
+            try {
+                value = convertType(value, type);
+            } catch (Exception e) {
+                throw new Exception(String.format("The value {%s} can't be converted to the type {%s}",
+                        value, type));
+            }
+        } else
+            value = null;
+        return value;
+    }
+
+    private long getTimeStamp(Object value, boolean needInitFormatter, DateTimeFormatter format) throws Exception{
+        if (format == null && needInitFormatter) {
+            format = initFormatter((String) value);
+            if (format == null) {
+                throw new Exception(String.format("%s The time format not supported", value));
+            }
+        }
+
+        if (needInitFormatter) {
+            return Timestamp
+                    .valueOf(LocalDateTime.parse((String) value, format))
+                    .getTime();
+        } else {
+            return (Long) value;
+        }
+    }
```

Comment:
See note on timestamp formatting, based on Record Reader column parsing, it should not be necessary to parse strings to a `long`.

Reply:
Refine the code and remove all timestamp parser code